django/db/migrations/graph.py

```diff
@@ -32,26 +80,37 @@ class MigrationGraph(object):
     """
 
     def __init__(self):
+        self.node_map = {}
         self.nodes = {}
-        self.dependencies = {}
-        self.dependents = {}
+        self.cached = False
 
-    def add_node(self, node, implementation):
+    def add_node(self, key, implementation):
+        node = Node(key)
+        self.node_map[key] = node
         self.nodes[node] = implementation
```

Comment:
On second thought, there is a lot of external code that depends on `graph.nodes[node]` being a `Migration`. It requires over a dozen changes and in my opinion unnecessarily clutters external code. Maintaining both `node_map` and `nodes` alleviates that problem, and allows quick access to both the node and the migration, based on either the node or the key, at minimal cost.